File: src/unitInputDemand.jl
"""
    unitInputDemand(xT::AbstractArray{<:Real}, θ::Real, κ::Real, z::Real, αVec::AbstractArray{<:Real}, skipParamChecks::Bool = false) -> AbstractArray{<:Real}

Calculates unit labor demands given blueprint scale `θ`, blueprint shape `κ`, productivity `z`, an array of comparative advantage values `αVec` with H elements (one for each worker type), and an array `xT` of H-1 thresholds in task space.

# Arguments
- `xT`: An array of H-1 thresholds in task space.
- `θ`: Blueprint scale parameter.
- `κ`: Blueprint shape parameter.
- `z`: Productivity parameter.
- `αVec`: An array of comparative advantage values with H elements.
- `skipParamChecks`: A boolean indicating whether to skip parameter checks (default is false).

# Returns
- An array representing the labor demand for each labor type.
"""
function unitInputDemand(xT::AbstractArray{<:Real}, θ::Real, κ::Real, z::Real, αVec::AbstractArray{<:Real},  skipParamChecks::Bool = false)
    
    if !skipParamChecks
        @assert θ > 0
        @assert κ > 0
        @assert z > 0
        @assert all(diff(αVec) .> 0) # Alpha values are increasing
        @assert all(diff(xT) .>= 0) # Threshold values are non-decreasing
    end

    H = length(xT) + 1
    labor_input = fill(0.0, H, 1)
    xT = vcat(0.0, xT, Inf) # Add lowest and highest thresholds
    for h in 1:H
        α = αVec[h]
        Υ = α + 1/θ
        if Υ > 0.0
            labor_input[h] = component_positive_ups(Υ, κ, xT[h], xT[h+1]) / Υ^κ
        elseif Υ == 0.0
            labor_input[h] = (xT[h+1]^κ - xT[h]^κ) / (κ * gamma(κ))
        else
            labor_input[h] = component_negative_ups(Υ, κ, xT[h], xT[h+1])
        end
    end
    labor_input = labor_input ./ (z * θ^κ)
end

File: TaskBasedProduction/+TaskBasedProduction/unitInputDemand.m
function l = unitInputDemand(theta, kappa, z, alphaVec, xT, skipParamChecks)
    % unitInputDemand Calculates unit labor demands
    % 
    % Calculates unit labor demands given blueprint scale theta, blueprint shape kappa,
    % productivity z, an array of comparative advantage values alphaVec with H elements
    % (one for each worker type), and an array xT of H-1 thresholds in task space.
    %
    % Inputs:
    %   theta - blueprint scale (scalar)
    %   kappa - blueprint shape (scalar)
    %   z - productivity (scalar)
    %   alphaVec - array of comparative advantage values (vector)
    %   xT - array of thresholds in task space (vector)
    %   skipParamChecks - flag to skip parameter checks (boolean)
    %
    % Output:
    %   l - unit labor demands (vector)

    % Assign default value to skipParamChecks if not provided
    if nargin < 6
        skipParamChecks = false;
    end

    if ~skipParamChecks
        assert(theta > 0, 'theta must be greater than 0');
        assert(kappa > 0, 'kappa must be greater than 0');
        assert(z > 0, 'z must be greater than 0');
        assert(all(diff(alphaVec) > 0), 'alphaVec values must be increasing');
        assert(all(diff(xT) >= 0), 'xT values must be non-decreasing');
    end

    H = length(xT) + 1;
    l = zeros(H, 1);
    xT = [0.0; xT(:); Inf]; % Add lowest and highest thresholds

    for h = 1:H
        alpha = alphaVec(h);
        upsilon = alpha + 1/theta;
        if upsilon > 0.0
            l(h) = TaskBasedProduction.component_positive_ups(upsilon, kappa, xT(h), xT(h+1)) / upsilon^kappa;
        elseif upsilon == 0.0
            l(h) = (xT(h+1)^kappa - xT(h)^kappa) / (kappa * gamma(kappa));
        else
            l(h) = TaskBasedProduction.component_negative_ups(upsilon, kappa, xT(h), xT(h+1));
        end
    end
    l = l / (z * theta^kappa);
end
